core/src/main/java/com/graphhopper/routing/HeadingResolver.java

```diff
@@ -87,9 +97,18 @@ public static boolean isHeadingNearlyParallel(EdgeIteratorState edgeState, doubl
         if (closestPoint >= 0) {
             double fromLat = points.getLat(closestPoint - 1), fromLon = points.getLon(closestPoint - 1);
             double toLat = points.getLat(closestPoint), toLon = points.getLon(closestPoint);
-            double orientation = AngleCalc.ANGLE_CALC.calcOrientation(fromLat, fromLon, toLat, toLon);
+            double orientation = parallel
+                    ? AngleCalc.ANGLE_CALC.calcOrientation(fromLat, fromLon, toLat, toLon)
+                    : AngleCalc.ANGLE_CALC.calcOrientation(toLat, toLon, fromLat, fromLon);
             orientation = AngleCalc.ANGLE_CALC.alignOrientation(xAxisAngle, orientation);
-            return Math.abs(orientation - xAxisAngle) < Math.toRadians(angleDifference);
+            if (Math.abs(orientation - xAxisAngle) < Math.toRadians(angleDifference))
+                return true;
+
+            if (parallel && directedEdgeFilter.accept(edgeState.detach(true))) {
+                orientation = AngleCalc.ANGLE_CALC.calcOrientation(toLat, toLon, fromLat, fromLon);
+                orientation = AngleCalc.ANGLE_CALC.alignOrientation(xAxisAngle, orientation);
+                return Math.abs(orientation - xAxisAngle) < Math.toRadians(angleDifference);
+            }
         }
```

Comment:
@easbar can you have a look if this sounds better to you? (now the distance calculation needs to be done once and the angle calculation too ... still in rare cases the opposite angle calculation is required, see test)

Reply:
Hm, `isHeadingParallel` looks really complicated? And there is no test (that just tests this method) either. I still wonder why there is no method that just calculates the angle of the geometry near the query point. It would be immediately clear what this method does and it would be really easy to test as well.

`isHeadingNearlyParallel` on the other is hard to understand (for me). Like, what is this directedEdgeFilter argument supposed to do? For me this method just does too many things and therefore I find it hard to understand. Maybe this is more complicated than I thought and I am missing some details, will send a PR with my suggestion.

Reply:
(Fix via #2411)